### ISSUE TITLE
Avoid caching of remote queries for block tags

### ISSUE THREAD

**Issue by Wodann** [NomicFoundation/rethnet#88]
**Problem**

The [`send_request_body`](https://github.com/NomicFoundation/hardhat/blob/rethnet/main/crates/rethnet_eth/src/remote/client.rs#L132) function uses an on-disk caching mechanism that uses a hash of the request body as index into the cache.

When the request body queries a block tag (e.g. latest), this pins the returned value to the latest block at call-time. Future calls to latest will return that particular (historic) block.

**Expected behaviour**

Ideally, calls that contain non-deterministic input data should be cached using a resolved, deterministic value.

Alternatively, we can avoid doing cache-lookups for non-deterministic input data.

**Potential Solution**

Before storing a response in cache, resolve any non-deterministic data in the request query, before caching.

**Definition of Done**
- The cache operates as expected

NOTE: This might require a refactor to cache something other than JSON queries

**Comment by alcuadrado** [comment]
Small note about this: 

It's not just `latest` that we should avoid caching. We shouldn't cache the queries to the state of the last N blocks of the chain. 

What should N be? That depends on the network, so we should derive it from the chainId. A static map with a few known chainIds and a default should be enough for now.

**Comment by alcuadrado** [comment]
In a sense, what I just described was a way to force the cache results to be deterministic if we are using block numbers or tags as `BlockSpec`. 

There's the possibility of using block hashes, but (1) getting the block hash may mean an extra request, and (2) it's unclear how widely supported blocktags with block hashes are.

**Comment by Wodann** [comment]
`eth_blockNumber` is another example of a method that should never be cached, as it returns the latest block number

**Comment by agostbiro** [comment]
`earliest`, `finalized` and `safe` are to be resolved to actual block numbers to save it into the cache

**Comment by agostbiro** [comment]
While [implementing](https://github.com/NomicFoundation/hardhat/pull/4319) this, the following questions came up:

- If the block [safety threshold](https://github.com/NomicFoundation/hardhat/blob/caa504fe0e53c183578f42d66f4740b8ec147051/packages/hardhat-core/src/internal/hardhat-network/provider/utils/reorgs-protection.ts) of 5 block depth is correct? I thought "safe" should be one epoch or 32 blocks.
- If there are methods other than [eth_getBlockByNumber](https://www.quicknode.com/docs/ethereum/eth_getBlockByNumber) where we can resolve the block number for a non-determinstic block spec argument (e.g. "safe") from the result? This is the only one I found.
- If we need the symbolic block spec resolution for batch requests? It's quite challenging to have a generic solution there as we have opaque result types in the batch call handler. We currently don't make batch requests that include `eth_getBlockByNumber`.

**Comment by alcuadrado** [comment]
> While [implementing](https://github.com/NomicFoundation/hardhat/pull/4319) this, the following questions came up:
> 
> * If the block [safety threshold](https://github.com/NomicFoundation/hardhat/blob/caa504fe0e53c183578f42d66f4740b8ec147051/packages/hardhat-core/src/internal/hardhat-network/provider/utils/reorgs-protection.ts) of 5 block depth is correct? I thought "safe" should be one epoch or 32 blocks.

That's interesting. 32 is definitely playing safe. I think for this use case it can make sense to use it.

5 was an unjustified guess.

> * If there are methods other than [eth_getBlockByNumber](https://www.quicknode.com/docs/ethereum/eth_getBlockByNumber) where we can resolve the block number for a non-determinstic block spec argument (e.g. "safe") from the result? This is the only one I found.

Note sure if I understand the question. There are other rpc methods that accept blocktags, like `eth_call` and `eth_estimateGas`.


> * If we need the symbolic block spec resolution for batch requests? It's quite challenging to have a generic solution there as we have opaque result types in the batch call handler. We currently don't make batch requests that include `eth_getBlockByNumber`.

The JS implementation of this allows you to pass a callback to decide if something should be cached or not. Not sure if that's expressible in rust.

**Comment by Wodann** [comment]
If it is too complicated, feel free to just disable caching for block tags for now. It's more important to merge a fix for it caching at all than to reduce false negatives (i.e. not caching something that could be cached)

**Comment by agostbiro** [comment]
> That's interesting. 32 is definitely playing safe. I think for this use case it can make sense to use it.

Ok thanks, [changed it](https://github.com/NomicFoundation/hardhat/pull/4319/commits/c0c973e7d3489a723c02cbe9a9419c307a3731da) to 32 where it was 5 before.

> Note sure if I understand the question. There are other rpc methods that accept blocktags, like eth_call and eth_estimateGas.

The goal is here to store as much as possible in the cache. for example, If there is call such as `eth_getBlockByNumber("finalized", false)`, then we cannot serve this from the cache, because "finalized" is non-deterministic. 

Let's assume this call returns a block with number "0xc5043f". Now we can store the response as if it was a `eth_getBlockByNumber("0xc5043f", false)` call and in the future if there is a call to get this specific block by number, we can serve it from cache.

My question was if there are methods other than `eth_getBlockByNumber` where we can do this, because the resolved block number is returned in the response? (I don't think so.)

> The JS implementation of this allows you to pass a callback to decide if something should be cached or not. Not sure if that's expressible in rust.

The difficulty with Rust here is that batch calls can include different method calls and a specific ordering of heterogeneous types in a sequence that can be arbitrarily long is very difficult to express in Rust. 

I realized since asking this question that I was overthinking in a bit and at a cost of a small performance hit there is a simple solution which I implemented [here.](https://github.com/NomicFoundation/hardhat/pull/4319/commits/50643e3b474885b4319ad560ef7c9908754dea1e)  The reason it's not very performant is the opaque argument to resolver that requires the resolver to at least partially parse the argument. 

In any case, we don't have to solve this now, because we don't use the resolver with batch calls currently.
